### DISCUSSION TITLE
Proposal: New Stories and Controls API that leverages ViewComponent Previews

### DISCUSSION THREAD

**Discussion by jonspalmer** [jonspalmer/view_component_storybook#150] [Ideas]
Since this gem's inception there has been a steady background of issues that have been hard to resolve. Most center around the way Controls are declared and the fact that declaring Stories is different in style and feel from ViewComponent Previews.

Examples of these challenges:
* #70 
* #73  
* #127
* #128 
* #129

The challenges stem from the somewhat convoluted DSL like approach I initially took to controls.

### Proposal 1: `ViewComponent::Storybook::Stories` become `ViewComponent::Preview`s 

Specifically `class ViewComponent::Storybook::Stories < ViewComponent::Preview`
This allows us to take advantage of all the features of `ViewComponent::Preview` - helper tags, custom templates, multiple components etc.

Suppose we have a `ButtonComponent`:

```ruby
# app/components/button_component/rb
class ButtonComponent < ViewComponent::Base
  def initialize(button_text:)
    super
    @button_text = button_text
  end
end
```

Today you might declare stories like this

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
   story :ok do
     constructor(
       button_text: "OK"
     )
  end
end
```

In the new API we simply render a Preview:

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
  def ok
    render ButtonComponent.new(button_text: "OK")
  end
end
```

### Proposal 2: Decouple Control declaration from how and where they're used

Rather than having `constructor`, `content` and `slot` controls you simply declare controls outside of the story with names that correspond to the story's method params.

Today a constructor control is added like this:

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
   story :ok do
     constructor(
       button_text: text("OK")
     )
  end
end
```

Under the proposal you would declare the controls like this:

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
  controls do
    text(:button_text: "OK")
  end

  def ok(button_text: "OK")
    render ButtonComponent.new(button_text: button_text)
  end
end
```

This approach is simple, powerful and flexible. For example suppose you want to combine two controls in the content of a component. Previously this would require a custom control now we simply declare two controls and use them in the render:

```ruby
# test/components/stories/greeting_component_stories.rb

class GreetingComponentStories < ViewComponent::Storybook::Stories
  controls do
    text(:greeting: "Hi")
    text(:name: "Jane")
  end

  def default(greeting: , name:)
    render GreetingComponent.new do
      "#{greeting} #{name}"
    end
  end
end
```

Similarly its easy to see how controls could. be wired up to constructor params, content blocks, slots (polymorphic or not) and a host of other use cases.

### Open Question: How should Controls be declared

Key considerations for the Controls API are:
1. Declaring controls for individual stories
2. Declaring default values for controls

#### Option 1: Comment Annotations
Inspired by [Logbooks approach](https://lookbook.build/guide/previews/params/) controls could be declared using custom Yard tags

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
  # @control button_text text
  def ok(button_text: "OK")
    render ButtonComponent.new(button_text: button_text)
  end

  # @control button_text text
  def close(button_text: "Close")
    render ButtonComponent.new(button_text: button_text)
  end
end
```

**Pros**
* Relatively simple 
* Read's default value from the story parameter default

**Cons**
* Using comments to drive run time behavior isn't a common Ruby idiom
* Complex control arguments are tough to pass in comment and brittle. Logbook solves this by allowing you to reference a separate method of yml file
* Retrieving the default value is a little janky, it only supports primitives and relies of `instance_eval`
* Implementation is relatively complicated and requires YARD and custom tags

#### Option 2: Controls helper
Inspired by Rails controller filters 

```ruby
# test/components/stories/button_component_stories.rb

class ButtonComponentStories < ViewComponent::Storybook::Stories
  controls(only: :ok) do
    text(:button_text, "OK")
  end

  def ok(button_text: )
    render ButtonComponent.new(button_text: button_text)
  end

  controls(only: :close) do
    text(:button_text, "Close")
  end

  def close(button_text:)
    render ButtonComponent.new(button_text: button_text)
  end
end
```

**Pros**
* Easy to specify complicated control options - it's just code
* More Railsy
* Controls can be declared for all examples
* Simple internal implementation
* Similar api could be used to define parameters and layout for specific 

**Cons**
* Default values are set in the controls vs in the story method - doesn't feel quite like Previews

**Comment by TastyPi** [comment]
For option 2 of defining controls, would it be possible to have the default values be the default values of the preview methods? E.g.

```ruby
controls do
  text(:button_text)
end

def ok(button_text: "OK")
  render Button component.new(button_text:)
end

def cancel(button_text: "Cancel")
  render Button component.new(button_text:)
end
```

**Reply by jonspalmer** [reply]
@TastyPi I think that would be ideal. It's remarkably hard to implement and relatively janky. Default parameters are evaluated at method call time so are effectively impossible to inspect via normal techniques see https://bugs.ruby-lang.org/issues/8629

Instead the technique Logbook uses is
1. Use something like YARD to parse the source
2. Find the methods definition in the parsed source
3. Find the keyword param in the definition and `instance_eval` what you find

It's a bunch of code and it doesn't support all parameter default values. However, the API would be a lot better. 
In practice the most common pattern is to group Stories about the same component and vary the inputs. Having to declare controls for each story just to change the default would be a pain. If we fix that the API is also much nicer than the alternative approach in that you can declare the controls for all stories once.

**Comment by elia** [comment]
@jonspalmer I think the macros are the ones that make more sense for the controls, maybe would be cool to remove the `controls` block and look more _Railsy_, I guess any dynamic part could be moved to a lambda if necessary:

```rb
class ButtonComponentStories < ViewComponent::Storybook::Stories
  control :button_text, as: :text, default: "OK"
  # or
  control :button_text, as: :text, default: -> { "OK #{rand}" }, only: :rand_ok

  def ok(button_text: )
    render ButtonComponent.new(button_text: button_text)
  end

  def rand_ok(button_text: )
    render ButtonComponent.new(button_text: button_text)
  end
end
```

**Reply by jonspalmer** [reply]
@elia I like the simplicity of your suggested macros. certainly very _Railsy_. Using a lamba for the dynamic part is a nice option, though anything dynamic like that is always going to get evaluate at story json writing time vs render time so its not really that 'dynamic'. The example you gave would actually work as a argument to the story method. 

The problem I was thinking about for using the `controls` block was for complex stories with lots of controls is to group them and only have to declare the `only` block once. Maybe its

```ruby
class ButtonComponentStories < ViewComponent::Storybook::Stories
  control :button_text, as: :text, default: "OK"
  control :button_color, as: :color, default: :red
  
 with_control_opts(only: :rand_ok) do
   control :button_text, as: :text, default: -> { "OK #{rand}" }
   control :button_color, as: :color, default: :green
 end
  
  def ok(button_text: , button_color: )
    render ButtonComponent.new(button_text: button_text)
  end

  def rand_ok(button_text: button_color: )
    render ButtonComponent.new(button_text: button_text)
  end
end
```


Maybe that's a feature set we learn if anyone needs later? I suspect the most common use case for `only` is to change the defaults. If we can fix that with the story method default values the rest might be less important.

**Reply by elia** [reply]
We can surely wait and see if there's demand for the more complex scenario, I think at least _some_ cases can be handled from within the method, possibly using `**kwargs` and no default values
